fast64_internal/f3d/f3d_gbi.py

```diff
@@ -2158,35 +2183,71 @@ def getCurrentAreaKey(self, material):
             return self.area_data[self.current_area_index].makeKey()
 
 
+class FImageKey:
+    def __init__(
+        self, image: bpy.types.Image, texFormat: str, palFormat: str, imagesSharingPalette: list[bpy.types.Image] = []
+    ):
+        self.image = image
+        self.texFormat = texFormat
+        self.palFormat = palFormat
+        self.imagesSharingPalette = tuple(imagesSharingPalette)
+
+    def __hash__(self) -> int:
+        return hash((self.image, self.texFormat, self.palFormat, self.imagesSharingPalette))
+
+    def __eq__(self, __o: object) -> bool:
+        if not isinstance(__o, FImageKey):
+            return False
+        return (
+            self.image == __o.image
+            and self.texFormat == __o.texFormat
+            and self.palFormat == __o.palFormat
+            and self.imagesSharingPalette == __o.imagesSharingPalette
+        )
+
+
+class FPaletteKey:
+    def __init__(self, palFormat: str, imagesSharingPalette: list[bpy.types.Image] = []):
+        self.palFormat = palFormat
+        self.imagesSharingPalette = tuple(imagesSharingPalette)
+
+    def __hash__(self) -> int:
+        return hash((self.palFormat, self.imagesSharingPalette))
+
+    def __eq__(self, __o: object) -> bool:
+        if not isinstance(__o, FPaletteKey):
+            return False
+        return self.palFormat == __o.palFormat and self.imagesSharingPalette == __o.imagesSharingPalette
+
+
 class FModel:
-    def __init__(self, f3dType, isHWv1, name, DLFormat, matWriteMethod):
+    def __init__(self, f3dType: F3D, isHWv1: bool, name: str, DLFormat: "DLFormat", matWriteMethod: GfxMatWriteMethod):
         self.name = name  # used for texture prefixing
         # dict of light name : Lights
-        self.lights = {}
+        self.lights: TypedDict[str, Lights] = {}
         # dict of (texture, (texture format, palette format)) : FImage
-        self.textures = {}
+        self.textures: TypedDict[Union[FImageKey, FPaletteKey] : FImage] = {}
```

Comment:
```suggestion
        self.textures: TypedDict[Union[FImageKey, FPaletteKey], FImage] = {}
```
also afaict this isn't what `TypedDict` is for https://peps.python.org/pep-0589/ (just use `dict` ?)

Reply:
Whoops, didn't realize. I'll edit that.